### PR TITLE
Bug 2061062: ensure healthy quorum before config update

### PR DESCRIPTION
**Pull request by tjungblu**
/hold

still testing and writing tests

### CODE REVIEW

**Comment by tjungblu** [comment]
so it seems this is not enough, the static pods still seem to update even though the quorum error log is printed. Maybe it's a race between the controllers, but it's odd that the revision bumps without the env var controller.

**Comment by tjungblu** [comment]
okay getting close, there's another controller which doesn't check for quorum:

```
I0705 14:58:12.857518       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-etcd-operator", Name:"etcd-operator", UID:"57ddbcd6-1708-4911-87f8-d0cb0c89de52", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'SecretUpdated' Updated Secret/etcd-all-certs -n openshift-etcd because it changed
I0705 14:58:12.963827       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-etcd-operator", Name:"etcd-operator", UID:"57ddbcd6-1708-4911-87f8-d0cb0c89de52", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RevisionTriggered' new revision 8 triggered by "secret/etcd-all-certs has changed"
```

**Comment by tjungblu** [comment]
okay, still not there yet. Continues to restart the pods - will try again tomorrow and see. Logs just got lost through a container reboot :(

**Comment by tjungblu** [comment]
> I0706 10:32:05.204979       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-etcd-operator", Name:"etcd-operator", UID:"27a063e8-87a5-45e1-a06a-689ef8cba15a", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RevisionTriggered' new revision 9 triggered by "configmap/etcd-endpoints has changed"

hope that's the last one. I'll refactor this into a dedicated controller now.

**Comment by tjungblu** [comment]
/test e2e-aws-single-node
--


<br class="Apple-interchange-newline">

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+				u.FakeEtcdMemberWithoutServer(0),
+				u.FakeEtcdMemberWithoutServer(2),
+			},
+			etcdMembersEnvVar: "1,3",
+			expectedErr:       fmt.Errorf("skipping TargetConfigController reconciliation due to insufficient quorum"),
+		},
+		{
+			name: "Quorum not fault tolerant but bootstrapping",
+			objects: []runtime.Object{
+				u.BootstrapConfigMap(u.WithBootstrapStatus("mot complete")),
```
Comment (Elbehery) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go:77]: `not complete`

**Comment by tjungblu** [comment]
/hold 

AWS was doing great, but just testing this with vsphere and it wrecks the cluster. I've seen the node come up with a different IP address, that triggered the envvar init container to crashloop, which for some reason caused the whole cluster to become unreachable.

**Comment by tjungblu** [comment]
seems like my refactoring broke something, I'll leave this open for now as we need to fix the vertical scaling tests first
/hold

**Comment by tjungblu** [comment]
actually I think I'll have to refactor once more. We already have a similar utility in https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/ceohelpers/bootstrap.go

It also has couple better strategies for singlenode, delayed HA etc.

**Comment by tjungblu** [comment]
/retest-required

still to test:
[ ] figure out why vsphere does not update on re-addition of the same machine
[ ] bare metal
[ ] SNO

**Comment by tjungblu** [comment]
/test e2e-aws-single-node
--


<br class="Apple-interchange-newline">

**Comment by tjungblu** [comment]
SNO generally seems to have trouble here:

> EtcdEndpointsDegraded: EtcdEndpointsController can't evaluate whether quorum is safe: etcd cluster has quorum of 1 which is not fault tolerant: [{Member:ID:196205077345930339 name:"ip-10-0-188-82.us-west-2.compute.internal" peerURLs:"https://10.0.188.82:2380/" clientURLs:"https://10.0.188.82:2379/" Healthy:true Took:462.474µs Error:<nil>}]

just created another commit to ensure we're skipping the health checks in the SNO scenario.

**Comment by tjungblu** [comment]
finally, today I'm going to test baremetal and vsphere again on the same-node addition. But I'm confident this is good to go now :)

**Comment by tjungblu** [comment]
still couldn't get my hands on a BM cluster, but vsphere works like a charm now :heavy_check_mark:.

**Comment by tjungblu** [comment]
/hold cancel

trying to get a BM cluster in the meantime

**Review by Elbehery — COMMENTED** [review]
Very good work 👍🏽 🎉

File: pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjungblu) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller_test.go]: thanks!

File: pkg/operator/etcdcertsigner/etcdcertsignercontroller.go [hunk 2]
```diff
@@ -106,7 +109,16 @@ func NewEtcdCertSignerController(
 }
 
 func (c *EtcdCertSignerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
-	err := c.syncAllMasters(ctx, syncCtx.Recorder())
+	safe, err := c.quorumChecker.IsSafeToUpdateRevision()
```
[the remark this answers is not in the document]
Comment (tjungblu) [reply pkg/operator/etcdcertsigner/etcdcertsignercontroller.go:112]: yeah that's it, everything that could trigger a new revision is front-loaded with this check

**Comment by tjungblu** [comment]
bare metal looks good so far, I'm doing another run later today. I think I also need to create a regression test for this whole story.

File: pkg/operator/ceohelpers/bootstrap_test.go [hunk 3]
```diff
@@ -333,15 +363,19 @@ func Test_CheckSafeToScaleCluster(t *testing.T) {
...
+				fakeInfraStructure,
+				fakeEtcdClient)
 
-			if test.expectError != nil && actualErr == nil {
-				t.Errorf("expected error=%v, got %v", test.expectError, actualErr)
-			}
-			if test.expectError == nil && actualErr != nil {
-				t.Errorf("expected error=%v, got %v", test.expectError, actualErr)
-			}
+			assert.Equal(t, test.expectError, actualErr)
```
Comment (hasbro17) [inline pkg/operator/ceohelpers/bootstrap_test.go:378]: Nice, I was confused at the behavior before this until I realized we were never checking the error.

**Review by hasbro17 — APPROVED** [review]
/approve

**Pull request merged**